File: plugins/spec-kit-copilot-wizard/extensions/speckit-wizard-canvas/generation/compiler.mjs
// Compile Wizard pipeline composition into a standalone workflow blueprint.
import {
    CORE_CAPABILITIES,
    canonicalArgumentGuidance,
    canonicalDescription,
    canonicalLabel,
    isCanonical,
} from "../pipeline/canonical.mjs";
import {
    effectivePipelinePhases,
    stripCommandsPrefix,
} from "../pipeline/effective-phases.mjs";
import { assessVisualizationApplicability } from "./applicability.mjs";
import { validateWorkflowPaths } from "./generated-canvas-template/workspace-files.mjs";
import { commandViews } from "./generated-canvas-template/ui/command-views.mjs";

export class BlueprintValidationError extends Error {
    constructor(errors) {
        super(errors.map((entry) => entry.message).join("; "));
        this.name = "BlueprintValidationError";
        this.code = "BLUEPRINT_INVALID";
        this.errors = errors;
    }
}

function commandLookup(snapshot) {
    const lookup = new Map();
    for (const command of snapshot?.commands ?? []) {
        for (const candidate of [command?.id, command?.commandName]) {
            const id = stripCommandsPrefix(candidate);
            if (typeof id !== "string" || !id) continue;
            lookup.set(id, command);
            if (id.startsWith("speckit.") && isCanonical(id.slice("speckit.".length))) {
                lookup.set(id.slice("speckit.".length), command);
            }
        }
    }
    return lookup;
}

function compositionCommand(snapshot, normalizedId) {
    return (snapshot?.composition?.artifacts ?? []).find((entry) => (
        entry?.kind === "command"
        && stripCommandsPrefix(entry.id) === normalizedId
    )) ?? null;
}

function normalizedCommandName(id, command) {
    const explicit = stripCommandsPrefix(command?.commandName);
    if (typeof explicit === "string" && explicit) {
        if (explicit.startsWith("speckit.")) return explicit;
        if (isCanonical(explicit)) return `speckit.${explicit}`;
        return explicit;
    }
    const normalized = stripCommandsPrefix(id);
    if (isCanonical(normalized)) return `speckit.${normalized}`;
    return normalized;
}

export function skillNameForCommand(commandName) {
    if (typeof commandName !== "string") return null;
    const normalized = commandName.trim().toLowerCase();
    if (!/^speckit(?:\.[a-z0-9][a-z0-9_-]*)+$/.test(normalized)) return null;
    return normalized.replace(/\./g, "-").replace(/_/g, "-");
}

function artifactFor(snapshot, originalId, normalizedId, command, allowCoreFallback = true) {
    return command?.artifactTemplatePath
        ?? snapshot?.phases?.[originalId]?.artifactTemplatePath
        ?? snapshot?.phases?.[`commands/${normalizedId}`]?.artifactTemplatePath
        ?? command?.artifact
        ?? command?.artifactPath
        ?? snapshot?.phases?.[originalId]?.artifactPath
        ?? snapshot?.phases?.[`commands/${normalizedId}`]?.artifactPath
        ?? (allowCoreFallback ? CORE_CAPABILITIES[`speckit.${normalizedId}`]?.writesTo : null)
        ?? null;
}

function sourceFor(snapshot, normalizedId, command, canonical) {
    const artifact = (snapshot?.composition?.artifacts ?? []).find((entry) => (
        entry?.id === `commands/${normalizedId}` || entry?.id === normalizedId
    ));
    const active = (artifact?.stack ?? []).find((layer) => layer?.active) ?? artifact?.stack?.[0] ?? null;
    if (active?.layer === "preset" || active?.layer === "extension") {
        return {
            kind: active.layer,
            id: active.extensionId ?? active.presetId ?? null,
            version: active.version ?? null,
            skillPath: active.sourcePath ?? null,
        };
    }
    const sourceText = String(command?.source ?? (canonical ? "core" : "extension"));
    const [kind, id] = sourceText.includes(":") ? sourceText.split(":", 2) : [sourceText, null];
    return {
        kind: kind || (canonical ? "core" : "extension"),
        id: id || null,
        version: null,
        skillPath: null,
    };
}

const CONTRIBUTION_ID_RE = /^[a-z0-9][a-z0-9._-]*$/i;

function catalogSource(item) {
    const sourceName = typeof item?.source === "string" ? item.source : null;
    const downloadUrl = typeof item?.downloadUrl === "string" && /^https:\/\//i.test(item.downloadUrl)
        ? item.downloadUrl
        : null;
    if (downloadUrl) {
        try {
            const parsed = new URL(downloadUrl);
            if (!parsed.hostname || parsed.username || parsed.password || /[\r\n]/.test(downloadUrl)) return null;
        } catch {
            return null;
        }
        return { name: sourceName ?? "default", url: downloadUrl, direct: true };
    }
    return null;
}

function contributionRecords(snapshot, kind, steps, errors, requireInstallationApproval) {
    const catalogItems = kind === "preset"
        ? snapshot?.catalog?.presets
        : snapshot?.catalog?.extensions;
    const compositionItems = kind === "preset"
        ? snapshot?.composition?.presets
        : snapshot?.composition?.extensions;
    const activeProviders = new Set(
        steps
            .filter((step) => step.source.kind === kind && step.source.id)
            .map((step) => step.source.id),
    );
    const compositionById = new Map(
        (Array.isArray(compositionItems) ? compositionItems : [])
            .filter((item) => item?.id)
            .map((item) => [String(item.id), item]),
    );
    const compositionOrder = new Map(
        (Array.isArray(compositionItems) ? compositionItems : [])
            .filter((item) => item?.id)
            .map((item, index) => [String(item.id), index]),
    );
    const records = new Map();
    for (const item of Array.isArray(catalogItems) ? catalogItems : []) {
        if (item?.active !== true) continue;
        const id = String(item.installedId ?? item.id ?? "");
        if (!CONTRIBUTION_ID_RE.test(id)) {
            errors.push({
                code: "setup_contribution_invalid",
                path: `catalog.${kind}s`,
                message: `Installed ${kind} "${id || "(missing id)"}" has no portable id.`,
            });
            continue;
        }
        const composed = compositionById.get(id) ?? compositionById.get(String(item.id ?? ""));
        const source = catalogSource(item);
        const priority = Number.isInteger(item.priority)
            ? item.priority
            : (Number.isInteger(composed?.priority) ? composed.priority : null);
        const precedence = Number.isInteger(item.cliOrder)
            ? item.cliOrder
            : (compositionOrder.get(id) ?? null);
        const sourceName = typeof item?.source === "string" ? item.source : null;
        const builtInSource = sourceName == null || sourceName === "default" || sourceName === "community";
        const invalidDownload = item.downloadUrl != null && item.downloadUrl !== "" && !source;
        if (invalidDownload || ((!builtInSource || (requireInstallationApproval && sourceName === "community")) && !source)) {
            errors.push({
                code: "setup_contribution_source_missing",
                path: `catalog.${kind}s`,
                message: `Installed ${kind} "${id}" from source "${sourceName ?? "default"}" has no portable HTTPS install URL.`,
            });
            continue;
        }
        records.set(id, {
            kind,
            id,
            enabled: typeof item.enabled === "boolean"
                ? item.enabled
                : (composed ? composed.enabled !== false : (activeProviders.has(id) ? true : null)),
            ...(priority !== null ? { priority } : {}),
            ...(precedence !== null ? { precedence } : {}),
            ...(source ? { source } : {}),
        });
    }
    for (const item of compositionById.values()) {
        const id = String(item.id ?? "");
        if (!CONTRIBUTION_ID_RE.test(id)) continue;
        if (!records.has(id)) {
            records.set(id, {
                kind,
                id,
                enabled: item.enabled !== false,
                ...(Number.isInteger(item.priority) ? { priority: item.priority } : {}),
                ...(compositionOrder.has(id) ? { precedence: compositionOrder.get(id) } : {}),
            });
        }
    }
    for (const id of activeProviders) {
        if (!CONTRIBUTION_ID_RE.test(id)) {
            errors.push({
                code: "setup_provider_invalid",
                message: `Selected ${kind} provider "${id}" has no portable id.`,
            });
            continue;
        }
        records.set(id, { ...(records.get(id) ?? { kind, id }), enabled: true });
    }
    return [...records.values()].sort((left, right) => {
        const leftOrder = Number.isInteger(left.precedence) ? left.precedence : Number.MAX_SAFE_INTEGER;
        const rightOrder = Number.isInteger(right.precedence) ? right.precedence : Number.MAX_SAFE_INTEGER;
        return leftOrder - rightOrder || left.id.localeCompare(right.id);
    });
}

function requiredSkills(steps, errors) {
    const records = new Map();
    for (const step of steps) {
        const provider = {
            kind: step.source.kind,
            id: step.source.id ?? null,
        };
        if (provider.kind !== "core" && !provider.id) {
            errors.push({
                code: "skill_provider_unknown",
                path: `pipeline[${step.index}]`,
                message: `Selected skill "${step.skillName}" has no identifiable ${provider.kind} provider.`,
            });
            continue;
        }
        if (!records.has(step.skillName)) {
            records.set(step.skillName, {
                name: step.skillName,
                invocation: step.invocation,
                commandName: step.commandName,
                provider,
            });
        }
    }
    return [...records.values()].sort((left, right) => left.name.localeCompare(right.name));
}

function phaseHints(snapshot, originalId, normalizedId) {
    return snapshot?.phases?.[originalId]
        ?? snapshot?.phases?.[`commands/${normalizedId}`]
        ?? snapshot?.phases?.[normalizedId]
        ?? {};
}

function phaseLabel(value) {
    const label = String(value ?? "").trim();
    return label ? label.charAt(0).toUpperCase() + label.slice(1) : label;
}

const TRANSIENT_COMPLETION = new Set([
    "speckit.analyze",
    "speckit.implement",
    "speckit.taskstoissues",
]);

export function compileBlueprint(snapshot, metadata, options = {}) {
    const errors = [];
    const warnings = [];
    if (options.requireInstallationApproval !== undefined && typeof options.requireInstallationApproval !== "boolean") {
        errors.push({
            code: "installation_approval_invalid",
            path: "requireInstallationApproval",
            message: "Require installation approval must be a boolean.",
        });
    }
    const requireInstallationApproval = options.requireInstallationApproval === true;
    const phases = effectivePipelinePhases(snapshot);
    const lookup = commandLookup(snapshot);
    const steps = [];

    if (!phases.length) {
        errors.push({ code: "pipeline_empty", message: "The effective pipeline must contain at least one command." });
    }

    for (let index = 0; index < phases.length; index += 1) {
        const entry = phases[index];
        const originalId = entry?.id;
        const normalizedId = stripCommandsPrefix(originalId);
        if (typeof normalizedId !== "string" || !normalizedId) {
            errors.push({ code: "phase_id_invalid", path: `pipeline[${index}].id`, message: `Pipeline step ${index + 1} has no valid command id.` });
            continue;
        }
        const canonical = isCanonical(normalizedId);
        const command = lookup.get(normalizedId) ?? null;
        const composedCommand = compositionCommand(snapshot, normalizedId);
        if (!canonical && !command && !composedCommand) {
            errors.push({ code: "command_unknown", path: `pipeline[${index}].id`, message: `Pipeline command "${normalizedId}" is not present in the snapshot command registry.` });
            continue;
        }
        const commandName = normalizedCommandName(normalizedId, command);
        const skillName = skillNameForCommand(commandName);
        if (!skillName) {
            errors.push({ code: "skill_mapping_invalid", path: `pipeline[${index}].id`, message: `Command "${commandName}" cannot be mapped to a Copilot skill.` });
            continue;
        }
        const label = phaseLabel(command?.shortLabel
            ?? command?.title
            ?? (canonical ? canonicalLabel(normalizedId) : commandName.split(".").at(-1)));
        const source = sourceFor(snapshot, normalizedId, command, canonical);
        const artifactPath = artifactFor(snapshot, originalId, normalizedId, command,
            commandName !== "speckit.constitution" || source.kind === "core");
        const hints = phaseHints(snapshot, originalId, normalizedId);
        const canonicalGuidance = canonicalArgumentGuidance(normalizedId);
        const argumentGuidance = {
            hint: typeof hints?.argsHint === "string" && hints.argsHint.trim()
                ? hints.argsHint
                : (canonical ? canonicalGuidance.hint : ""),
            whenEmpty: typeof hints?.argsWhenEmpty === "string" && hints.argsWhenEmpty.trim()
                ? hints.argsWhenEmpty
                : (canonical ? canonicalGuidance.whenEmpty : ""),
        };
        const persistent = Boolean(artifactPath) && !TRANSIENT_COMPLETION.has(commandName);
        if (!artifactPath) {
            warnings.push({
                code: "artifact_unknown",
                path: `pipeline[${index}]`,
                message: `No artifact target is known for "${commandName}"; the generated canvas must show it as a transient phase.`,
            });
        }
        steps.push({
            index,
            instanceKey: `${index}:${normalizedId}`,
            id: normalizedId,
            commandName,
            skillName,
            invocation: `/skill:${skillName}`,
            label: String(label ?? commandName),
            description: String(command?.helpText ?? command?.description ?? composedCommand?.description ?? hints?.description ?? (canonical ? canonicalDescription(normalizedId) : "")),
            source,
            artifact: {
                pathTemplate: artifactPath,
                persistent,
                completionSignal: persistent ? "artifact" : "transient",
            },
            arguments: {
                hint: argumentGuidance.hint,
                whenEmpty: argumentGuidance.whenEmpty,
            },
            optional: command?.optional === true,
            predecessors: index === 0 ? [] : [index - 1],
        });
    }

    if (errors.length) throw new BlueprintValidationError(errors);
    const constitutions = steps.filter((step) => step.commandName === "speckit.constitution");
    const projectArtifacts = constitutions.length
        ? { constitution: { instanceKey: constitutions[0].instanceKey, required: true } }
        : undefined;
    let views;
    try {
        views = commandViews({ pipeline: { steps }, projectArtifacts });
    } catch (error) {
        throw new BlueprintValidationError([{ code: "constitution_contract_unsupported", message: error.message }]);
    }
    const applicability = assessVisualizationApplicability(steps, views.workflow);
    if (!applicability.ok) throw new BlueprintValidationError(applicability.errors);
    try {
        validateWorkflowPaths({ pipeline: { steps }, projectArtifacts, runtime: { itemRoot: applicability.itemRoot } });
    } catch (error) {
        throw new BlueprintValidationError([{ code: "workflow_path_invalid", path: "pipeline", message: error.message }]);
    }
    const skills = requiredSkills(steps, errors);
    const presets = contributionRecords(snapshot, "preset", steps, errors, requireInstallationApproval);
    const extensions = contributionRecords(snapshot, "extension", steps, errors, requireInstallationApproval);
    if (errors.length) throw new BlueprintValidationError(errors);
    return {
        schemaVersion: 2,
        kind: "speckit-wizard-linear-canvas",
        metadata: {
            extensionId: metadata.extensionId,
            displayName: metadata.displayName,
            description: metadata.description,
            workflowListName: metadata.workflowListName ?? "Workflows",
        },
        pipeline: {
            topology: "linear",
            steps,
        },
        ...(projectArtifacts ? { projectArtifacts } : {}),
        setup: {
            requiresSpecKit: true,
            requireInstallationApproval,
            integration: {
                id: "copilot",
                skillsMode: true,
            },
            requiredSkills: skills,
            presets,
            extensions,
        },
        runtime: {
            visualStyle: "spec-kit-wizard",
            workflowMode: applicability.workflowMode,
            itemRoot: applicability.itemRoot,
            userProvidesSlug: options.userProvidesSlug === true,
            multiInstance: applicability.workflowMode === "item",
            supportsArtifactPreview: true,
            supportsRerun: true,
            supportsSse: true,
            requiredActions: ["list_items", "setup_workflow", "run_phase"],
        },
        warnings,
    };
}

File: spec-kit-extensions/pipeline-canvas-generator/scripts/lib/compiler.py
"""Compile a request-bound linear blueprint without reading live composition."""

from copy import deepcopy

from request import validate_request


DEFAULT_ARGUMENTS = {
    "hint": "Provide guidance to focus or scope this phase.",
    "whenEmpty": "If left empty, the phase will run with default behavior using the existing artifacts.",
}
CORE_ARGUMENTS = {
    "constitution": (
        "Enter your project's governing principles and development guidelines that will guide all subsequent development.",
        "If left empty, a new constitution will be drafted from your repo context (README, docs) for review; otherwise, an existing constitution will be changed.",
    ),
    "specify": (
        "Describe what you want to build - focus on the what and why, not the tech stack.",
        "A description is required.",
    ),
    "clarify": (
        "Provide areas of concern to focus clarification pass.",
        "If left empty, the full spec will be scanned across categories of impact areas (scope, data model, UX, integration, etc.).",
    ),
    "plan": (
        "Provide your tech stack and architecture choices.",
        "If left empty, the plan will be derived from the spec.md and constitution.md alone, marking missing technical decisions as needing clarification.",
    ),
    "tasks": (
        "Add guidance for task generation like groupings, priorities, and areas to emphasize.",
        "If left empty, a full, dependency-ordered tasks.md will be generated directly from plan.md and spec.md (with constitution.md as governing constraints).",
    ),
    "implement": (
        "Add guidance for the implementation.",
        "If left empty, all tasks in tasks.md will be implemented in dependency order, updating progress markers as each completes.",
    ),
    "analyze": (
        "Add a specific concern for analysis to focus on.",
        "If left empty, a full consistency and quality analysis will be performed across spec.md, plan.md, and tasks.md (with constitution.md as governing authority).",
    ),
    "taskstoissues": (
        "Add issue-creation guidance like labels, milestone, and assignees.",
        "If left empty, one GitHub issue per task will be created in tasks.md on the current repo's git remote.",
    ),
}


def _source(snapshot: dict, command: dict) -> dict:
    contributor = next(
        (layer for layer in command["stack"]
         if layer["active"] and layer["layer"] in ("project", "preset", "extension")),
        None,
    )
    if contributor is None:
        return {"kind": "core", "id": None, "version": None, "skillPath": None}
    if contributor["layer"] == "project":
        return {
            "kind": "project", "id": "project", "version": None,
            "skillPath": contributor["sourcePath"],
        }
    provider = next(
        entry for entry in snapshot["providers"]
        if entry["kind"] == contributor["layer"] and entry["id"] == contributor["sourceId"]
    )
    return {
        "kind": provider["kind"], "id": provider["id"],
        "version": provider["version"], "skillPath": contributor["sourcePath"],
    }


def _setup_requirements(
    snapshot: dict, commands: list[dict], steps: list[dict],
    skill_hashes: dict | None = None,
) -> dict:
    providers = {}
    precedence = {"preset": {}, "extension": {}}
    for provider in snapshot["providers"]:
        kind = provider["kind"]
        precedence[kind][provider["id"]] = len(precedence[kind])
    for command in commands:
        for layer in command["stack"]:
            if not layer["active"] or layer["layer"] not in ("preset", "extension"):
                continue
            kind, package_id = layer["layer"], layer["sourceId"]
            source = next(
                entry for entry in snapshot["providers"]
                if entry["kind"] == kind and entry["id"] == package_id
            )
            providers[(kind, package_id)] = {
                "kind": kind, "id": package_id, "version": source["version"],
                "enabled": source["enabled"], "priority": source["priority"],
                "precedence": precedence[kind][package_id],
                "installedSource": deepcopy(source["source"]),
                "manifestPath": source["manifestPath"],
                **({"manifestSha256": source["manifestSha256"]}
                   if "manifestSha256" in source else {}),
            }
    skills = {}
    for step in steps:
        skills[step["skillName"]] = {
            "name": step["skillName"], "invocation": step["invocation"],
            "commandName": step["commandName"],
            "provider": {key: step["source"][key] for key in ("kind", "id")},
            **({"sha256": skill_hashes[step["skillName"]]}
               if skill_hashes and skill_hashes.get(step["skillName"]) else {}),
        }
    return {
        "requiresSpecKit": True,
        "requireInstallationApproval": False,
        "compositionFingerprint": snapshot["compositionFingerprint"],
        "integration": {"id": "copilot", "skillsMode": True},
        "requiredSkills": [skills[name] for name in sorted(skills)],
        "presets": sorted(
            (value for (kind, _), value in providers.items() if kind == "preset"),
            key=lambda value: value["precedence"],
        ),
        "extensions": sorted(
            (value for (kind, _), value in providers.items() if kind == "extension"),
            key=lambda value: value["precedence"],
        ),
    }


def compile_blueprint(request: dict) -> dict:
    """Preserve phase order, bound outputs and Specify provider provenance."""
    validate_request(request)
    workflow = request["workflow"]
    instance_configuration = request["instanceConfiguration"]
    snapshot = workflow["artifactSnapshot"]
    commands_by_name = {
        row["name"]: row for row in snapshot["artifacts"] if row["kind"] == "command"
    }
    commands = [commands_by_name[phase] for phase in workflow["selectedPhases"]]
    steps = []
    item_roots = set()
    hint_roots = set()
    constitution = None
    for index, (phase, binding, command) in enumerate(
        zip(workflow["selectedPhases"], workflow["phaseOutputs"], commands, strict=True)
    ):
        result = binding["contract"]["result"]
        artifact_path = result.get("pathTemplate")
        if phase == "speckit.constitution":
            if result["kind"] != "artifact" or "<" in artifact_path or ">" in artifact_path:
                raise ValueError("Unsupported project Constitution output contract")
        elif artifact_path and "<slug>" in artifact_path:
            prefix = artifact_path.split("<slug>", 1)[0]
            if "<" in prefix or ">" in prefix:
                raise ValueError("Unsupported workflow artifact root")
            if result["kind"] == "artifact":
                item_roots.add(prefix + "<slug>")
            else:
                hint_roots.add(prefix + "<slug>")
        bare = phase.removeprefix("speckit.")
        skill = phase.replace(".", "-")
        hint, when_empty = CORE_ARGUMENTS.get(bare, (
            DEFAULT_ARGUMENTS["hint"], DEFAULT_ARGUMENTS["whenEmpty"],
        ))
        source = _source(snapshot, command)
        step = {
            "index": index,
            "instanceKey": f"{index}:{bare}",
            "id": bare,
            "commandName": phase,
            "skillName": skill,
            "invocation": f"/skill:{skill}",
            "label": bare.split(".")[-1].replace("-", " ").title(),
            "description": command.get("description", ""),
            "source": source,
            "artifact": {
                "pathTemplate": artifact_path,
                "persistent": result["kind"] == "artifact",
                "completionSignal": result["kind"],
            },
            "arguments": {"hint": hint, "whenEmpty": when_empty},
            "optional": False,
            "predecessors": [] if index == 0 else [index - 1],
        }
        if phase == "speckit.constitution":
            constitution = {"instanceKey": step["instanceKey"], "required": True}
        steps.append(step)
    if len(item_roots) > 1:
        raise ValueError("The pipeline uses multiple independent item roots")
    item_root = next(iter(item_roots), None)
    if item_root is None and len(hint_roots) == 1:
        item_root = next(iter(hint_roots))
    for step in steps:
        path = step["artifact"]["pathTemplate"]
        if step["artifact"]["completionSignal"] == "hint" and path and "<slug>" in path:
            if not item_root or not path.startswith(item_root + "/"):
                step["artifact"]["pathTemplate"] = None
    blueprint = {
        "schemaVersion": 2,
        "kind": "speckit-wizard-linear-canvas",
        "metadata": {
            "extensionId": request["canvas"]["id"],
            "displayName": request["canvas"]["displayName"],
            "description": request["canvas"]["description"],
            "workflowListName": request["canvas"]["workflowListName"],
        },
        "pipeline": {"topology": "linear", "steps": steps},
        "phaseOutputs": deepcopy(workflow["phaseOutputs"]),
        "setup": {
            **_setup_requirements(snapshot, commands, steps, workflow["requiredSkillHashes"]),
            "requireInstallationApproval": instance_configuration["installationMode"] == "prompt",
        },
        "runtime": {
            "visualStyle": "spec-kit-wizard",
            "workflowMode": "item" if item_root else "project",
            "itemRoot": item_root,
            "userProvidesSlug": instance_configuration["workflowSlug"]["userProvided"],
            "multiInstance": bool(item_root),
            "supportsArtifactPreview": True,
            "supportsRerun": True,
            "supportsSse": True,
            "requiredActions": ["list_items", "setup_workflow", "run_phase"],
        },
        "warnings": [],
    }
    if constitution is not None:
        blueprint["projectArtifacts"] = {"constitution": constitution}
    return blueprint
